kernel/work.c

```diff
@@ -0,0 +1,952 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * Second generation work queue implementation
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <wait_q.h>
+#include <spinlock.h>
+#include <errno.h>
+#include <sys/scheduler.h>
+#include <sys/printk.h>
+
+/* Lock to protect the internal state of all work items, and
+ * pending_cancels.
+ *
+ * No queue locks may be held when this is taken, but a queue lock may
+ * be taken while this is held.
+ */
+static struct k_spinlock lock;
+
+/* Invoked by work thread */
+static void handle_flush(struct k_work *work)
+{
+	struct z_work_flusher *flusher
+		= CONTAINER_OF(work, struct z_work_flusher, work);
+
+	k_sem_give(&flusher->sem);
+}
+
+static inline void init_flusher(struct z_work_flusher *flusher)
+{
+	k_sem_init(&flusher->sem, 0, 1);
+	k_work_init(&flusher->work, handle_flush);
+}
+
+/* List of pending cancellations. */
+static sys_slist_t pending_cancels;
+
+/* Initialize a canceler record and add it to the list of pending
+ * cancels.
+ *
+ * Invoked with work lock held.
+ *
+ * @param canceler the structure used to notify a waiting process.
+ * @param work the work structure that is to be canceled
+ */
+static inline void init_work_cancel(struct z_work_canceller *canceler,
+				    struct k_work *work)
+{
+	k_sem_init(&canceler->sem, 0, 1);
+	canceler->work = work;
+	sys_slist_append(&pending_cancels, &canceler->node);
+}
+
+/* Complete cancellation of a work item and unlock held lock.
+ *
+ * Invoked with work lock held.
+ *
+ * Invoked from a work queue thread.
+ *
+ * Releases work lock before returning.
+ *
+ * Reschedules.
+ *
+ * @param work the work structre that has completed cancellation
+ * @param key the key to the held lock
+ */
+static void _finalize_cancel_and_unlock(struct k_work *work,
+					k_spinlock_key_t key)
+{
+	struct z_work_canceller *wc, *tmp;
+	sys_slist_t cancels;
+
+	/* Clear this first, so released high-priority threads don't
+	 * see it when doing things.
+	 */
+	atomic_clear_bit(&work->flags, K_WORK_CANCELING_BIT);
+
+	/* Move all pending cancels associated with this work item to
+	 * another list, under lock.  (We can't signal the release
+	 * while the lock is held because it's a reschedule point,
+	 * which would release the lock.)
+	 */
+	sys_slist_init(&cancels);
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&pending_cancels, wc, tmp, node) {
+		if (wc->work == work) {
+			sys_slist_remove(&pending_cancels, NULL, &wc->node);
+			k_sem_give(&wc->sem);
+		}
+	}
+
+	/* Release the lock.  No need to yield here, the work thread
+	 * does that.
+	 */
+	k_spin_unlock(&lock, key);
```

Comment:
hmm. generally when implementing critical sections it's better to have the lock/unlock operations be in the same function scope instead of split like this. I would strongly prefer this be moved.